Space Dragon Rework and Dynamic Ruleset

itsmeow opened the pull request:
## About The Pull Request

Ports:
- https://github.com/tgstation/tgstation/pull/49344 (**big**)
- https://github.com/tgstation/tgstation/pull/50196
- https://github.com/tgstation/tgstation/pull/50932
- https://github.com/tgstation/tgstation/pull/53944
- https://github.com/tgstation/tgstation/pull/54124 (**medium**)
- https://github.com/tgstation/tgstation/pull/54774 (just the space dragon fix)
- https://github.com/tgstation/tgstation/pull/55759
- https://github.com/tgstation/tgstation/pull/57428 (just the IDEA, heavily modified since our code is completely different)
- https://github.com/tgstation/tgstation/pull/57942 (**medium**, gust is not a right-click, since we don't have that)
- https://github.com/tgstation/tgstation/pull/58392
- https://github.com/tgstation/tgstation/pull/60058
- https://github.com/tgstation/tgstation/pull/60259 (**big nerfs**)
- https://github.com/tgstation/tgstation/pull/65311 (thanks ivan)
- https://github.com/tgstation/tgstation/pull/67630
- https://github.com/tgstation/tgstation/pull/68988
- https://github.com/tgstation/tgstation/pull/69359
- https://github.com/tgstation/tgstation/pull/70522

Includes additional dynamic tweaks to the space dragon midround that is added in 49344 from https://github.com/tgstation/tgstation/pull/65997 and https://github.com/tgstation/tgstation/pull/68571

See original PRs for changes.

Includes some additional changes:
- Had to rework the ritual claw which used tail sweep, now it just uses the xeno one instead

## Why It's Good For The Game

The old space dragon is really bad and just has to gib a random head, then be killed by everyone or gib random people to survive. Bad murderbone weird space dragon. This gives it actual unique objectives that are interesting and fun to fight.

It also adds it to dynamic, providing a greater variety of midrounds.

## Testing Photographs and Procedure
<details>
<summary>Screenshots&Videos</summary>

https://user-images.githubusercontent.com/10366817/184504878-a7acfb56-7a7e-4d1c-8476-a7f759466258.mp4

![image](https://user-images.githubusercontent.com/10366817/213823001-206567ca-e826-44b1-bfa9-ec5a5254625a.png)

</details>

## Changelog
:cl: itsmeow, Indie-ana Jones, redguy999, Timberpoes, Ryll-Ryll, Ghommie, read-0nly, ivanmixo, JohnFulpWillard, tralezab
tweak: Space Dragon now has its own role preference. Be sure to enable it!
refactor: The Space Dragon has totally new objectives where it can spawn rifts that summon sentient space carp, and must defend the rifts until three are created, flooding the station with sentient carp.
add: Space Dragon is now a ruleset in dynamic as well as a station event.
tweak: Space Dragons can now set their color.
add: Space Dragon now has its own death sound.
tweak: Space Dragon roar is now quieter.
/:cl:

IndieanaJones committed:
[READY] Space Dragon Rework (#49344)

* Remove Space Dragon from drake.dm

* Add space_dragon.dm

* Update spacedragon.dmi

* Add actions_space_dragon.dmi

* Add carp_rift.dmi

* Load space_dragon.dm

* Update space_dragon.dm antagonist file

* Update spawn_dragon.dm Spawn Event

* Update misc.dm for Space Dragon Win Definition

* Update role_preferences.dm to include Space Dragon

* Update ticker.dm to include Space Dragon news message

* First Trailing Newline for Travis

* Second Trailing Newline for Travis

* Changing rift to rifts

* Update antag file with proposed new weight and time

* Update the Greeting to be Correct

* Update space_dragon.dm mob file

* Remove Space Dragon Win Definition

* Remove Space Dragon News Message

* Undo Whitespace change

* Update space_dragon.dm

* Travis got mad about how I spelled CENTCOM in Autodoc Info

* Update code/modules/events/space_dragon.dm

Co-Authored-By: moo <11748095+ExcessiveUseOfCobblestone@users.noreply.github.com>

* Update space_dragon.dm

* Update space_dragon.dm

* Display Space Dragon in his own section of the Antag Panel

* Use the new Spacewalk trait

* Add Speed Boost After Rift Charge, Passive Healing While On Rift

* Adds Space Dragon to the Dynamic Midround System

* Notify Ghosts When a Carp Spawn is Added

* Actually Fix That

* Reduces Space Dragon's Spawn Weight from 10 to 5

Co-authored-by: moo <11748095+ExcessiveUseOfCobblestone@users.noreply.github.com>

IndieanaJones committed:
Fix carp rifts not actually being infinite (#50196)

Fixes carp rifts not spawning in infinite carps upon Space Dragon
completing his objective. Also renders the carp rift above things like
doors, as per request. It not already doing this was an oversight, and
gives Space Dragon winning some weight. Also makes it easier for ghosts
to use the rift if it shares a tile with something else.

redguy999 committed:
Space dragons are now large mobs (#50932)

you can mark queens and Praetorians, so why not a giant space dragon?

Timberpoes committed:
Minor refactor to space dragon code to fix various bugs and issues (#53944)

IndieanaJones committed:
Space Dragon QOL and Additional Features (#54124)

Co-authored-by: Jared-Fogle <35135081+Jared-Fogle@users.noreply.github.com>

IndieanaJones committed:
Space Dragon Fixes (#55759)

IndieanaJones committed:
[READY] Space Dragon Quality Improvement Update (#57942)

 qol: Space Dragon's gust attack is now tied to right click instead of an ability button
balance: Space Dragon has received numerous balance changes, most notably that Space Dragon will stall the shuttle until it loses or completes its objectives
fix: Fixed numerous bugs with Space Dragon, most notably the rifts not vanishing after Space Dragon's death
soundadd: Space Dragon now has a custom roar for dying and winning
imagedel: Deleted now unused Space Dragon ability icon
refactor: Space Dragon's speed changes are now controlled using movespeed modifiers instead of direct speed adjustments

Co-authored-by: Ghom <42542238+Ghommie@users.noreply.github.com>
Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

IndieanaJones committed:
Reduce Loudness of Space Dragon's Roar (#58392)

read-0nly committed:
Adds random number to space-dragon-rift-spawned space carp (#60058)

Co-authored-by: Watermelon914 <37270891+Watermelon914@users.noreply.github.com>

IndieanaJones committed:
Space Dragon Nerfs and Adjustments (#60259)

ivanmixo committed:
Give space dragon ability to make it's sprite small (#65311)

About The Pull Request

Allows space dragon to replace it's sprite with a carp sprite like the xeno queen can turn it's sprite into a drone.

This means only the dragon sees itself as small. Others still see it as giant.
Why It's Good For The Game

The space dragon sprite is pretty big, so you can't attack stuff directly north of you. This solves that.
Changelog

cl
qol: space dragon can now make it's sprite smaller
/cl

JohnFulpWillard committed:
Space dragon despawning empties their contents (#67630)

itsmeow committed:
Fix rift timer

itsmeow committed:
Grammar fix

itsmeow committed:
Fix linting issue

itsmeow committed:
Port a fix from tg/54774

itsmeow committed:
Space dragons can fly (tg/57428)

itsmeow committed:
Lower random event chance back to pre-rework chance

PowerfulBacon commented on code/modules/antagonists/space_dragon/space_dragon.dm:
```diff
@@ -1,38 +1,55 @@
-	to_chat(owner, "<b>Using Tail Sweep will let me get the better of those who come too close.</b>")
-	to_chat(owner, "<b>Attacking dead bodies will allow me to gib them to restore health.</b>")
-	to_chat(owner, "<b>From the wizard's writings, he had been studying this station and its hierarchy.  From this, I know who leads the station, and will kill them so the station underlings see me as their new leader.</b>")
+	to_chat(owner, "<b>Endless time and space we have moved through.  We do not remember from where we came, we do not know where we will go.  All space belongs to us.\n\
```
This string has double spaces in it

PowerfulBacon commented on code/modules/mob/living/simple_animal/hostile/space_dragon.dm:
```diff
@@ -0,0 +1,727 @@
+	newcarp.unique_name = TRUE
+	var/datum/antagonist/space_dragon/S = dragon?.mind?.has_antag_datum(/datum/antagonist/space_dragon)
+	if(S)
+		S.carp += newcarp.mind
```
Potential source of hard dels?

itsmeow replied:
minds cannot be deleted afaik, /datum/team tracks minds in the "members" list and has no del handling, so it should be fine?

PowerfulBacon commented on code/modules/mob/living/simple_animal/hostile/space_dragon.dm:
```diff
@@ -0,0 +1,727 @@
+		icon_state = "carp_rift_charged"
+		set_light_color(LIGHT_COLOR_YELLOW)
+		update_light()
+		armor = list(MELEE = 100, BULLET = 100, LASER = 100, ENERGY = 100, BOMB = 100, BIO = 100, RAD = 100, FIRE = 100, ACID = 100)
```
Whats the point of setting the armour if its indestructible anyway

PowerfulBacon commented on code/modules/mob/living/simple_animal/hostile/space_dragon.dm:
```diff
@@ -0,0 +1,727 @@
+		rift.dragon = null
+		rift_list -= rift
+		if(!QDELETED(rift))
+			QDEL_NULL(rift)
```
Why qdel null a local variable?

PowerfulBacon commented on code/modules/mob/living/simple_animal/hostile/space_dragon.dm:
```diff
@@ -0,0 +1,727 @@
+	playsound(src, 'sound/vehicles/rocketlaunch.ogg', 100, TRUE)
+	for(var/obj/structure/carp_rift/rift in rift_list)
+		rift.dragon = null
+		rift_list -= rift
```
Just use rift_list.cut() at the end since you are removing all of them anyway

PowerfulBacon commented on code/modules/mob/living/simple_animal/hostile/space_dragon.dm:
```diff
@@ -0,0 +1,727 @@
+	tiredness_mult = 5
+	playsound(src, 'sound/vehicles/rocketlaunch.ogg', 100, TRUE)
+	for(var/obj/structure/carp_rift/rift in rift_list)
+		rift.dragon = null
```
Might as well be in rift destroy()

PowerfulBacon commented on code/modules/mob/living/simple_animal/hostile/space_dragon.dm:
```diff
@@ -0,0 +1,727 @@
+ */
+/mob/living/simple_animal/hostile/space_dragon/proc/destroy_rifts()
+	rifts_charged = 0
+	add_movespeed_modifier(MOVESPEED_ID_DRAGON_DEPRESSION, multiplicative_slowdown = 5)
```
is this permanent?

itsmeow replied:
Yes, this is effectively the lose condition for a space dragon

PowerfulBacon commented on code/modules/mob/living/simple_animal/hostile/space_dragon.dm:
```diff
@@ -0,0 +1,727 @@
+	if(timer != 10)
+		pixel_y = pixel_y + 2;
+		addtimer(CALLBACK(src, .proc/useGust, timer + 1), 1.5)
+		return
```
Use animate instead of calling itself over and over again with a timer.
Also 1.5 seconds is a long time to be frozen for in order to use this ability, you will probably have lost half your health by then, or at least everyone will be out of the way.

itsmeow replied:
Changed to 1 second, plus there was no cooldown so I added one for 3 seconds.

PowerfulBacon commented on code/modules/mob/living/simple_animal/hostile/space_dragon.dm:
```diff
@@ -0,0 +1,727 @@
+		addtimer(CALLBACK(src, .proc/useGust, timer + 1), 1.5)
+		return
+	pixel_y = 0
+	icon_state = "spacedragon_gust_2"
```
If this is an animated icon, you can use flick() to play the animation once and then go back to the previous icon state.

itsmeow replied:
Not animated

PowerfulBacon commented on code/modules/mob/living/simple_animal/hostile/space_dragon.dm:
```diff
@@ -0,0 +1,727 @@
+	sound_to_playing_players('sound/creatures/space_dragon_roar.ogg')
+	for(var/obj/structure/carp_rift/rift in rift_list)
+		rift.carp_stored = 999999
+		rift.time_charged = rift.max_charge
```
Space dragon shouldn't be a round end antag with an immortal carp spawner. There should still be a way to recover

itsmeow replied:
It doesn't shuttle call anymore. The carps are also ghost spawn only so it's not effectively infinite.

itsmeow replied:
The station can choose to deal with the carp by containing the rift or killing the carp that come out.

PowerfulBacon commented on code/modules/mob/living/simple_animal/hostile/space_dragon.dm:
```diff
@@ -0,0 +1,727 @@
+ * - Can swallow mob corpses to heal for half their max health.  Any corpses swallowed are stored within him, and will be regurgitated on death.
+ * - Can tear through any type of wall.  This takes 4 seconds for most walls, and 12 seconds for reinforced walls.
+ */
+/mob/living/simple_animal/hostile/space_dragon
```
Delete the rifts on destroy

itsmeow replied:
Removed on death **if** their objective is not complete. Rifts are supposed to stay if they succeed and then die so this would break that.

PowerfulBacon commented on code/modules/mob/living/simple_animal/hostile/space_dragon.dm:
```diff
@@ -0,0 +1,727 @@
+		playsound(src, 'sound/effects/splat.ogg', 50, TRUE)
+		visible_message("<span class='danger'>[src] vomits up [consumed_mob]!</span>")
+		consumed_mob.forceMove(loc)
+		consumed_mob.Paralyze(50)
```
Is there not a revival signal we can check for rather than doing this in life?

PowerfulBacon commented on code/modules/mob/living/simple_animal/hostile/space_dragon.dm:
```diff
@@ -0,0 +1,727 @@
+		visible_message("<span class='danger'>[src] vomits up [consumed_mob]!</span>")
+		consumed_mob.forceMove(loc)
+		consumed_mob.Paralyze(50)
+	if((rifts_charged == 3 || (SSshuttle.emergency.mode == SHUTTLE_DOCKED && rifts_charged > 0)) && !objective_complete)
```
They victory if the shuttle docks? Seems a bit cheap

itsmeow replied:
TG upstream said this was because crew was just calling shuttle instead of dealing with the rifts. Lore wise, everyone left and at least one rift is open, so realistically, the dragon wins, as the station is now the dragon's. I could add a check for if the dragon is on station z level.

PowerfulBacon commented on code/modules/mob/living/simple_animal/hostile/space_dragon.dm:
```diff
@@ -0,0 +1,727 @@
+		destroy_rifts()
+		empty_contents()
+		playsound(src, 'sound/magic/demon_dies.ogg', 100, TRUE)
+		QDEL_NULL(src)
```
Just use a world.time based timer

PowerfulBacon commented on code/modules/mob/living/simple_animal/hostile/space_dragon.dm:
```diff
@@ -0,0 +1,727 @@
+		destroy_rifts()
+		empty_contents()
+		playsound(src, 'sound/magic/demon_dies.ogg', 100, TRUE)
+		QDEL_NULL(src)
```
Might be a better idea to limit their abilities until they place the rift rather than just delete them. Prevents just going in for murderbone while not outright deleting slow dragons

itsmeow replied:
Their abilities are already pretty minimal, I could remove fire breath but honestly you get 6 minutes to go onto station and make a rift. I could add the depression instead of deleting since that happens when you fail regularly.

itsmeow replied:
Removing the qdel and empty_contents. Dragon will basically get nerfed hard (destroy_rifts adds the depression and prevents future rifht placing) if they don't place rifts in time instead of just straight up deleted.

PowerfulBacon commented on code/modules/mob/living/simple_animal/hostile/space_dragon.dm:
```diff
@@ -0,0 +1,727 @@
+		var/obj/mecha/M = target
+		M.take_damage(50, BRUTE, "melee", 1)
+
+/mob/living/simple_animal/hostile/space_dragon/Move()
+	if(!using_special)
+		..()
```
Will this prevent all moves like shuttle movement?

itsmeow replied:
I have no idea

itsmeow replied:
Looks like hiero and megafauna dragon do the same thing by overriding `Move()` should be fine since z-level transitions usually do forceMove?

PowerfulBacon commented on code/modules/mob/living/simple_animal/hostile/space_dragon.dm:
```diff
@@ -0,0 +1,727 @@
+  * If the name is invalid, will re-prompt the dragon until a proper name is chosen.
+  */
+/mob/living/simple_animal/hostile/space_dragon/proc/dragon_name()
+	var/chosen_name = sanitize_name(reject_bad_text(stripped_input(src, "What would you like your name to be?", "Choose Your Name", real_name, MAX_NAME_LEN)))
+	if(!chosen_name)
+		to_chat(src, "<span class='warning'>Not a valid name, please try again.</span>")
+		dragon_name()
+		return
+	to_chat(src, "<span class='notice'>Your name is now <span class='name'>[chosen_name]</span>, the feared Space Dragon.</span>")
+	fully_replace_character_name(null, chosen_name)
```
This is probably a bad idea ngl

itsmeow replied:
ask admins, I'll remove if needed

itsmeow replied:
we've decided on a pre-selected lists of proper nouns and adjectives for names like
`Dragur the Terrible`. You will be able to pick from both lists so you can have a static but reasonable space dragon name.

itsmeow replied:
nevermind, too much work. commented out the line that calls the rename method. Leaving this in though

PowerfulBacon commented on code/modules/mob/living/simple_animal/hostile/space_dragon.dm:
```diff
@@ -0,0 +1,727 @@
+  * Clears the current overlay on space dragon and adds a proper one for whatever animation he's in.
+  */
+/mob/living/simple_animal/hostile/space_dragon/proc/add_dragon_overlay()
+	cut_overlays()
```
should be update_dragon_overlay

PowerfulBacon commented on code/modules/mob/living/simple_animal/hostile/space_dragon.dm:
```diff
@@ -0,0 +1,727 @@
+	if(!at)
+		return
+	var/angle = ATAN2(at.x - src.x, at.y - src.y) + offset
+	var/turf/T = get_turf(src)
+	for(var/i in 1 to range)
+		var/turf/check = locate(src.x + cos(angle) * i, src.y + sin(angle) * i, src.z)
+		if(!check)
+			break
+		T = check
```
Whats the point of this check?

PowerfulBacon replied:
If its to find the point that is range units away from the dragon, just use some basic trig to find it directly without needing to loop

itsmeow replied:
After reading the method doc I still have a very vague idea, but it's not for getting a specific spot x units away, it's for... creating a line of turfs? As an array? It's weird.

PowerfulBacon commented on code/modules/mob/living/simple_animal/hostile/space_dragon.dm:
```diff
@@ -0,0 +1,727 @@
+		if(L in hit_list)
+			continue
+		hit_list += L
+		L.adjustFireLoss(30)
```
No armour check?

itsmeow replied:
It never had this before in the old space dragon. I think armor is handled automatically? I can't find anywhere armor is used really.

PowerfulBacon commented on code/modules/mob/living/simple_animal/hostile/space_dragon.dm:
```diff
@@ -0,0 +1,727 @@
+  * * atom/movable/A - The thing being consumed
+  */
+/mob/living/simple_animal/hostile/space_dragon/proc/eat(atom/movable/A)
+	if(A && A.loc != src)
```
```suggestion
	if(A?.loc != src)
```

PowerfulBacon commented on code/modules/mob/living/simple_animal/hostile/space_dragon.dm:
```diff
@@ -0,0 +1,727 @@
+		return
+
+	// Can we increase the carp spawn pool size?
+	if(last_carp_inc >= carp_interval)
```
Use a world.time based timer

itsmeow replied:
This actually works better by adding delta time as it currently does, the number never gets larger than `carp_interval` and using `addtimers` doesn't really work well since the interval is dynamic. If the interval changes, this needs to respond instantly, `world.time` operates on game ticks and will not work well during lag.

I fixed the other one, but this one should stay as it is

PowerfulBacon reviewed (dismissed):
Space dragon should be a threat but not a guaranteed shuttle call at any point.

itsmeow committed:
General code improvement, fix overlay coloring being broken

itsmeow committed:
Make messages more accurate to new changes

itsmeow committed:
Rework gust animation, add cooldown, fix bug where overlay would appear on small sprite, remove renaming

tralezab committed:
Space Dragons can no longer place rifts on open spaces (#69359)

About The Pull Request

Dragons can no longer place rifts on open space, upper open space, etc. Basically, if it has a zlevel openly under it, the dragon can't do it

Why It's Good For The Game

Fighting the dragon rift should be about killing the dragon or killing the rift, not figuring out how to bridge out to it. Open space rifts will always be at a massive, massive advantage over ones on stable ground and I don't want that to be the biggest factor that goes into placing a rift.

itsmeow commented on code/modules/mob/living/simple_animal/hostile/space_dragon.dm:
```diff
@@ -0,0 +1,749 @@
+				RegisterSignal(L, COMSIG_LIVING_REVIVE, .proc/living_revive)
+				if(eat(L))
+					adjustHealth(-L.maxHealth * 0.25)
+			return
```
set is_swallowing = FALSE here

itsmeow committed:
Make carp immune to fireball and fix swallowing

itsmeow committed:
Merge branch 'master' into space-dragon-rework

itsmeow committed:
Refactor (tg/68988) and fix rifts being invincible

itsmeow committed:
Port most of tg/70522 with fixes

itsmeow committed:
Finish, adding wavespeak

itsmeow committed:
Update for Dynamic 2023
